File: Design Patterns/Structural Design Patterns/Composite Design Pattern/Implementation/composite_design_pattern_demo.py
#Component Interface that acts as blueprint for parent and child
class Employee:
    def get_id(self):
        pass 

    def get_name(self):
        pass

    def get_salary(self):
        pass

    def print(self):
        pass 

    def add(self, emp):
        pass

    def remove(self, emp):
        pass

    def getChild(self, index):
        pass

# Parent Class
class BankManager(Employee):
    def __init__(self, id, name, salary):
        self.__id = id 
        self.__name = name
        self.__salary = salary
        self.__emps = []
    
    def get_id(self):
        return self.__id
    
    def get_name(self):
        return self.__name
    
    def get_salary(self):
        return self.__salary

    def print(self):
        for emp in self.__emps:
            print(emp)
    
    def add(self, emp):
        self.__emps.append(emp)
    
    def remove(self, emp):
        self.__emps.remove(emp)
    
    def getChild(self, index):
        return self.__emps[index]


# Child Class
class Cashier(Employee):
    def __init__(self, id, name, salary):
        self.__id = id 
        self.__name = name
        self.__salary = salary
    
    def get_id(self):
        return self.__id

    def get_name(self):
        return self.__name
    
    def get_salary(self):
        return self.__salary
    
    def add(self, emp):
        raise NotImplementedError('Operation not supported for child')
    
    def remove(self, emp):
        raise NotImplementedError('Operation not supported for child')
    
    def getChild(self, index):
        return None
    
    def __str__(self) :
        return f'Cashier id = {self.__id}, name = {self.__name}, salary = {self.__salary}'


emp1 = Cashier(1, 'Aravind reddy', 100000)
emp2 = Cashier(2, 'Pandi Anil', 600000)

manager = BankManager(3, 'Anil', 1000000)

manager.add(emp1)
manager.add(emp2)

manager.print()

File: Design Patterns/Structural Design Patterns/Composite Design Pattern/Example/CompositeDesignPatternDemo.java
import java.util.List;
import java.util.ArrayList;

interface Employee{
    Integer getId();
    String getName();
    Double getSalary();
    void print();
    void add(Employee emp);
    void  remove(Employee emp);
    Employee getChild(int i);
}

class BankManager implements Employee{

    private int id;
    private String name;
    private Double salary;
    List<Employee> emps = new ArrayList<>();

    public BankManager(int id, String name, Double salary){
        this.id = id;
        this.name = name;
        this.salary = salary;
    }

    @Override
    public Integer getId() {
        return id;
    }

    @Override
    public String getName() {
       return name;
    }

    @Override
    public Double getSalary() {
       return salary;
    }

    @Override
    public void print() {
        for(Employee emp : emps){
            System.out.println(emp);
        }
    }

    @Override
    public void add(Employee emp) {
        emps.add(emp);
    }

    @Override
    public void remove(Employee emp) {
       emps.remove(emp);
    }

    @Override
    public Employee getChild(int i) {
        return emps.get(i);
    }
    
}


class Cashier implements Employee{

    private Integer id;
    private String name;
    private Double salary;

    public Cashier(int i, String string, int j) {
        //TODO Auto-generated constructor stub
    }

    @Override
    public Integer getId() {
       return id;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public Double getSalary() {
        return salary;
    }

    @Override
    public void print() {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'print'");
    }

    @Override
    public void add(Employee emp) {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Add Operation not supported for Leaf");
    }

    @Override
    public void remove(Employee emp) {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Not applicable");
    }

    @Override
    public Employee getChild(int i) {
        return null;
    }
}

public class CompositeDesignPatternDemo{
    public static void main(String[] args) {
        Employee emp1 = new Cashier(101, "Sonam Kumar",20000);
        Employee emp2 = new Cashier(102, "mOHAN KUMAR", 3000);
        Employee manager = new BankManager(103, "Champak Setty", 100000.0);

        manager.add(emp1);
        manager.add(emp2);

        manager.print();
    }
}